sycl/include/CL/sycl/detail/cstr.hpp

```diff
@@ -0,0 +1,151 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+///  \file
+///  This file contains meta programming utilities to manipulate string
+///  constants stored in a type like cstr<char, 'a', 'b'>.
+///  This is used for kernel properties because they must be constant and the
+///  API reflects that.
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef SYCL_DETAIL_CSTR_HPP
+#define SYCL_DETAIL_CSTR_HPP
+
+#include "CL/sycl/detail/pi.h"
+#include <tuple>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+
+namespace sycl {
+
+namespace detail {
+
+/// utility to make any functor's operator() be const even if it originally
+/// wasn't. This is useful to bypass the restriction that kernel operator() must
+/// be const.
+template <typename T> struct const_cheating_wrapper {
+  mutable std::decay_t<T> t;
+  template <typename... Ts> auto operator()(Ts &&... ts) const & {
+    return std::forward<T &>(t)(std::forward<Ts>(ts)...);
+  }
+  template <typename... Ts> auto operator()(Ts &&... ts) const && {
+    return std::forward<T &&>(t)(std::forward<Ts>(ts)...);
+  }
+};
+template <typename T> auto make_const_cheater(T &&t) {
+  return const_cheating_wrapper<T>{std::forward<T>(t)};
+}
+
+/// transform a type like: cstr<char, 'a', 'b'>
+/// into a string "ab"
+template <typename> struct StrPacker {};
+
+/// utility type to manipulate string in constant context
+template <typename CharT, CharT... charpack> struct cstr {
+  using Char = CharT;
+  static constexpr unsigned size = sizeof...(charpack);
+  /// access character at position Idx in the string
+  template <unsigned Idx>
+  using at = __type_pack_element<Idx, cstr<CharT, charpack>...>;
+};
+
+template <typename CharT, CharT... P1> struct StrPacker<cstr<CharT, P1...>> {
+  static constexpr const CharT Str[] = {P1...};
+};
+
+/// utility to merge string with a separator
+template <typename...> struct concat {};
+
+template <typename CharT, CharT... Sep, CharT... P1>
+struct concat<cstr<CharT, Sep...>, cstr<CharT, P1...>> {
+  using type = cstr<CharT, P1...>;
+};
+
+template <typename CharT, CharT... Sep, CharT... P1, CharT... P2,
+          typename... Rest>
+struct concat<cstr<CharT, Sep...>, cstr<CharT, P1...>, cstr<CharT, P2...>,
+              Rest...> {
+  using type =
+      typename concat<cstr<CharT, Sep...>, cstr<CharT, P1..., Sep..., P2...>,
+                      Rest...>::type;
+};
+
+template<typename...Ts>
+using concat_t = typename concat<Ts...>::type;
+
+#ifdef __SYCL_DEVICE_ONLY__
+
+/// We are using a compiler extention here and we don't want any warning about
+/// it so we suppress that warning.
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wgnu-string-literal-operator-template"
+
+/// utility to transform a string literal into its cstr form.
```

Comment:
```suggestion
/// Utility to transform a string literal into its cstr form.
```
Start sentences with uppercase!